feat: update landing page header

n05la3 opened the pull request:
- Style the header search form and add animations to scroll area
- Add secondary header that is swaped with the main header depending on user scroll position
- Add the docs landing page
- random code refactoring

n05la3 committed:
fix(landing-page): remove scaffolding section

n05la3 committed:
fix(landing-page): update styles to adhere to mockup

n05la3 committed:
refactor(landing-page): update landing page styles

n05la3 committed:
fix(landing page): issue which closes search result form before click on links in it are fired

IlCallo committed:
chore: enable eslint as formatter

IlCallo committed:
fix(landing): fix wrong options names and paths

IlCallo committed:
feat: fix many details, simplify code where possible, use semantic tags

IlCallo committed:
refactor: remove obsolete pages and sections

n05la3 committed:
feat(landing page): add secondary header nav items

n05la3 committed:
feat(landing page): update index page

n05la3 committed:
refactor(mainlayout): improve readability of 'stars' function

n05la3 committed:
feat(mainlayout): add shadow to dropdown menus

n05la3 committed:
fix(landing page): random fixes

n05la3 committed:
feat(landing page): add secondary header

n05la3 committed:
feat(landing page): merge with dev

IlCallo commented on docs/src/components/landing-page/MainLayoutHeader.vue:
```diff
@@ -1,119 +1,249 @@
+    leave-active-class="animated slideOutUp"
+  >
+    <q-header
+      v-if="scrollOffset < SWAP_HEADER_OFFSET"
```
The big header version should appear any time the user scrolls up at least a given delta (100px?) and reduce when the user scrolls down by the same delta.
Seeing it live actually makes me think it's better to have it this way, where it only expands when you reach top, but we still want to see how it comes out.

Keep this version aside somewhere in case we need to rollback, but implement the one going up-and-down whenever the user scrolls up and down
I can already see this will be a pain point for us 🙃

n05la3 replied:
Just to be sure I follow your last description of what I'm to implement, are you talking about something like the reveal prop of q-header? https://quasar.dev/layout/header-and-footer#reveal-property

IlCallo commented on docs/src/components/landing-page/MainLayoutHeader.vue:
```diff
@@ -1,119 +1,249 @@
+              nav-item-class="text-size-16"
+            />
+          </div>
+          <div ref="searchForm">
```
Why is an additional div needed? Can't you add this on the search component directly?

n05la3 replied:
The wrapper div is necessary since we use it to know the region which does not cause the form to close. It wraps the search input field and the icon. Since this hiding and display of the search form is not the business of the search component itself, I preferred to move the wrapper to the parent.

IlCallo commented on docs/src/components/landing-page/MainLayoutHeader.vue:
```diff
@@ -1,119 +1,249 @@
-              :target="navItem.href? '_blank':'_self'"
-              :color="setActivePrimaryNavColor($route.path, navItem.path)"
+        <div class="row items-center">
+          <div
```
When search is in "box" mode, you need to add a small margin between the last button and the search box

IlCallo commented on docs/src/components/landing-page/MainLayoutHeader.vue:
```diff
@@ -208,13 +334,23 @@ export default defineComponent({
+    onMounted(() => {
+      document.body.addEventListener('click', (e) => {
+        // if click is perform on any region out of the search form area then
+        // close the search or open leave it open otherwise
```
I think this comment must be fixed

IlCallo commented on docs/src/components/landing-page/SearchQuasarForm.vue:
```diff
@@ -0,0 +1,144 @@
+      enter-active-class="animated fadeIn"
+      leave-active-class="animated fadeOut"
+    >
+      <q-input
```
When in input form, we need to show the same "button like" icon suggesting that pressing `/` will bring the focus to the search bar, as we have on current docs

![image](https://user-images.githubusercontent.com/10036108/151201278-0e6817e5-1911-4e10-b72f-1804a604af9d.png)

On that regards, right that feature is broken and throwns an error, you may need to adapt the listener
You can find the code in `use-search` file, it should work both when the input or the icon are shown

IlCallo commented on docs/src/components/landing-page/TwitterShowcaseCards.vue:
```diff
@@ -165,6 +147,11 @@ export default defineComponent({
 
 <style lang="scss" scoped>
+/*
+ [1], [2]: needed to create a white card behind the twitter card, this card gets shown when moving to previous carousel
```
Numbers here should only be one per "feature", not one per line of code used: you're mapping lines to the comment, not the other way around

Here there shouldn't be `[2]`, and the line marked with `[2]` should be marked with `[1]`
Dunno if the concept is clear, I can re-explain it if needed

IlCallo commented on docs/src/pages/Landing.vue:
```diff
@@ -1,71 +1,63 @@
+    <h2
+      class="lp-heading lp-heading--medium q-mb-lg cards-title normal-line-height"
+    >Have a look at Our most used pages</h2>
+    <div class="cards-grid lp-mb--large">
```
I'm pretty sure the whole card must be clickable here, not just the button, but ask Nick to be sure

IlCallo reviewed (changes requested):
A couple more details: the Quasar logo right now doesn't open a new page if you open it with the middle button of the mouse

Seach results also won't open into a new tab, please integrate this PR to solve that issue: https://github.com/quasarframework/quasar/pull/10914

There are also many warnings in the console, check them out and see what are they about

Check out the header into mobile version:
- it should have a height of 64px when in this mode, as opposed to 92px of other viewports
- the ripple area of the hamburger icon on the left (the white thing which appears when you hover it) is stretched probably due to flex options of the parent 

![image](https://user-images.githubusercontent.com/10036108/151206176-a33eab9c-b60b-4502-a4c6-2d6e26e627cc.png)

n05la3 committed:
feat(landing page): improve accessibility of search result links

n05la3 committed:
feat(header): swap headers when scrolling up and down at some point in the page

IlCallo commented on docs/src/layouts/DocLayout.vue:
```diff
@@ -45,13 +45,14 @@ q-layout.doc-layout(view="hHh LpR lff", @scroll="onScroll")
     router-view
 
-  q-page-scroller(scroll-offset="150" :offset="[18, 18]")
+  q-page-scroller(:scroll-offset="150" :offset="[18, 18]")
```
Why wasn't the normal "scroll-offset" ok here?

n05la3 replied:
The scroll-offset prop is expecting a Number, not a String. It was the reason for the many console error logs.

IlCallo commented on docs/src/layouts/DocLayout.vue:
```diff
@@ -82,6 +82,8 @@ export default {
   setup () {
     const $q = useQuasar()
     const $route = useRoute()
+    const headerScrollData = ref(null)
```
`ref` automatically initialize as `undefined` if nothing is provided, is there a particular reason you initialized it to `null` here?

IlCallo replied:
Check this also into refs from "use-search"

n05la3 replied:
I usually prefer to set variables which will eventually hold objects to ```null``` instead of leaving them as ```undefined```. This case was really a situation of personal preference of ```null``` over ```undefined``` for initializing variables that hold objects. And this is because ```typeof null``` is ```object```. But I won't argue it has any special benefits or so. Just preference! and with consistency I find it more informing.

n05la3 replied:
Apart from the extra typing would there be some particular reason to avoid this which I may be missing? I have updated the rest of the code, will wait for your response to update this if necessary.

IlCallo replied:
I'd advise you against using `null` in pretty much every situation, the fact that `typeof null` is `object` is part of the problem: `null` has a ton of weird interactions in JS and is usually the cause for many subtle bugs

The only cases where I found null to be useful is where it is explicitly required by code you don't have control over: sending up/getting a null value into/from GraphQL (undefined values are stripped away), modeling responses from an API that actually returns and expects null, in general any process based on `JSON.stringify`.

In all other cases, it's better to avoid it altogether and convert any null values to undefined as soon as possible

---
I agree with you that it could be informing if it where a well-known style standard, but Vue explicitly suggest to avoid filling in with undefined new refs (and I'd extend this to null, as is better to avoid null altogether)

For sure the problem you want to solve with this convention is more present in JS codebases as TS ones enforce types anyway
I would avoid it into Dreamonkey projects, as we almost always use TS and we are used to avoid initialization if not needed

IlCallo commented on docs/src/layouts/doc-layout/use-search.js:
```diff
@@ -38,8 +37,7 @@ export default function useSearch (scope, $q, $route) {
   const searchInputRef = ref(null)
-
-  const $router = useRouter()
+  const focusByKbd = ref(false)
```
**Always** avoid abbreviation, use `focusByKeyboard` instead and change the related event name accordingly

IlCallo commented on docs/src/assets/landing-page/nav-items.js:
```diff
@@ -9,13 +9,128 @@ import {
+        },
+        {
+          label: 'Vite Plugin',
+          path: 'start/vite-plugin'
```
This brings to nowhere, we probably need to rebase on upstream as we miss this page in our version

IlCallo commented on docs/src/components/landing-page/MainLayoutHeader.vue:
```diff
@@ -1,119 +1,262 @@
+          icon="menu"
+          round
+          @click="$emit('update:modelValue', !modelValue)"/>
+        <a
```
Can this work with a `router-link` instead?
In other words, is there a way to make this work with `$router.push({name: 'home'})` style but keeping the "open in new tab" behaviour when clicking on the middle mouse button?
I think `router-link` with `:to="{name: 'home'}"` may automatically manage this, but I'm not sure

n05la3 replied:
Yes, it works with ```router-link```. I agree that it's a preferable solution since the one with ```a``` tag cannot work with named routes.

IlCallo commented on docs/src/components/landing-page/SearchQuasarForm.vue:
```diff
@@ -0,0 +1,159 @@
+        @focus="onSearchFocus"
+      >
+        <template #append>
+          <div
```
This must have the look & feel of the existing one on current docs
It must look as a "floating" raised grey button, mimiking the look & feel of the key of an old keyboard

IlCallo commented on docs/src/components/landing-page/MainLayoutHeader.vue:
```diff
@@ -125,26 +268,25 @@ export default defineComponent({
-
-      isSearchFieldActive.value = !isSearchFieldActive.value
-    }
+    const searchForm = ref(null)
```
Same questions as the other `ref` initialized with `null`

n05la3 replied:
For a similar reason as noted above.

IlCallo reviewed (changes requested):
There are some details which are still off, but almost ready for merge

n05la3 committed:
refactor(landing page): random refactoring to address readability and accessibility related issues

IlCallo commented on docs/src/components/landing-page/SearchQuasarForm.vue:
```diff
@@ -27,10 +27,11 @@
         <template #append>
-          <div
+          <!-- justify-end: a tweak to center / in kbd, since the size of / depends on the font, flex flex-center does not quite place it at the center-->
+          <kbd
```
I totally missed there was an ad-hoc tag for this :O
Nice!

IlCallo commented on docs/src/components/landing-page/SearchQuasarForm.vue:
```diff
@@ -113,8 +114,8 @@ export default {
-    watch(() => scope.focusByKbd.value, () => {
-      emit('focus-by-kbd', scope.focusByKbd.value)
+    watch(() => scope.focusByKeyboard.value, () => {
+      emit('focus-by-kbd', scope.focusByKeyboard.value)
```
Please, no more abbreviations :'(
Use `focus-by-keyboard` instead

n05la3 committed:
refactor(landing page):
- remove null initialization of refs
- change focus-by-kbd to focus-by-keyboard
